### ISSUE TITLE
RFC: Flattening Configuration YAML

### ISSUE THREAD

**Issue by tylerjl** [elastic/puppet-elasticsearch#633]
The module currently accepts configuration hashes of the form (for example) `{ 'path.data' => '/mnt' }` or `{'path' => { 'data' => '/mnt' } }`. This is good, since it allows for either fully-qualifying configuration options or passing in structured configs that could come from hiera, for example.

Because of the way Elasticsearch can consume configuration options, a yaml key can be be fully qualified to define a value that isn't necessarily an atom (string, integer, etc.) For example, translating the following dot-notation flattened hash into a fully nested hash:

```
---
shield.http.ssl: true
shield.http.ssl.client.auth: optional
```

Means that the `ssl` key is going to have to choose between either storing `true` or `{'client' => {'auth' => 'optional'}}`. One way around this is to detect colliding keys such that they get qualified:

```
shield:
  http:
    ssl: true
    ssl.client:
      auth: optional
```

However, that behavior is mixing dot-notated and a purely hash-based representation, which doesn't even have clearly defined rules (i.e. how deep the nesting goes, etc.)

I'd like to write out `elasticsearch.yml` with all configuration options in the dot-notated long form, which should completely eliminate any ambiguities down the road. It's more verbose, and it doesn't render the yaml as nested key/values, but qualifying at least **some** keys is just a necessity to accommodate some configuration options, so I'd rather keep it consistent rather than mixed. Keep in mind we can still keep `config =>` accepting hashes, we would just convert everything internally in the manifests after accepting the hashes. This also means merging hashes (i.e. what comes from the defaults in the elasticsearch class) becomes much simpler.

I haven't started on this yet, but if there is any input for/against, please comment. I'll try and leave this up for a while in case there are any objections to due problems this could cause downstream for users.

**Comment by drnr53** [comment]
Currently running into this issue myself with the ssl example cited above, no objections here as our current non-controlled yml configs are in dot-notated form. Also, any temporary fix recommendations are welcome.

**Comment by tylerjl** [comment]
For all those interested, I've got a branch cooking in CI for this, [you can see the diff here](https://github.com/tylerjl/puppet-elasticsearch/compare/master...fully-qualified-config-yml). The user-facing side of this (for Puppet) should be the same - the only change is how the configuration hashes are transformed internally and written in the configuration file. Behavior in regards to merging array values should behave the same (when duplicate hash keys with array values are encountered, they are concatenated). 

In addition, because all configuration hashes are flattened and then merged with `concat_merge`, deep merging should be resolved (#456).

All `elasticsearch.yml` files will change due to this, which is kind of a big shift, but it should eliminate ambiguities in regards to how merging and concatenation behave. Fortunately the incoming changes in #655 will mitigate the surprise here and give people more control about when Elasticsearch restarts (and it won't, by default, after the configuration file changes due to this).

If anyone has feedback regarding the behavior in the aforementioned branch diff, please to comment on the relevant commit or provide feedback here.

**Comment by tylerjl** [comment]
Change implemented in #656.

Note that the only difference to the aforementioned ^ comment is that I pulled the `concat_merge` call that concatenated arrays together, as this wasn't the way it behaved before, so I'm erring on the side of old behavior for now. If it ever becomes a need for people, we can always just change the `merge` call if needed (the puppet function is still there).
